skydb/skydb.go

```diff
@@ -0,0 +1,159 @@
+		return nil, 0, ErrNotFound
+	}
+	if err != nil {
+		return nil, 0, errors.AddContext(err, "failed to read from registry")
+	}
+	b, err := db.Client.SkynetSkylinkGet(s.String())
+	if err != nil {
+		return nil, 0, errors.AddContext(err, "failed to download data from Skynet")
+	}
+	return b, rev, nil
+}
+
+// Write stores the given `data` in SkyDB under the given key set.
+func (db SkyDB) Write(data []byte, dataKey crypto.Hash, rev uint64) error {
+	skylink, err := uploadData(db.Client, data)
+	if err != nil {
+		return errors.AddContext(err, "failed to upload data")
+	}
+	_, err = registryWrite(db.Client, skylink, db.sk, db.pk, dataKey, rev)
+	if err != nil {
+		return errors.AddContext(err, "failed to write to the registry")
+	}
+	return nil
+}
+
+// registryWrite updates the registry entry with the given dataKey to contain the
+// given skylink. Returns a SkylinkV2.
+func registryWrite(c *client.Client, skylink string, sk crypto.SecretKey, pk crypto.PublicKey, dataKey crypto.Hash, rev uint64) (skymodules.Skylink, error) {
+	var sl skymodules.Skylink
+	err := sl.LoadString(skylink)
+	if err != nil {
+		return skymodules.Skylink{}, errors.AddContext(err, "failed to load skylink data")
+	}
+	// Update the registry with that link.
+	spk := types.SiaPublicKey{
+		Algorithm: types.SignatureEd25519,
+		Key:       pk[:],
+	}
+	srv := modules.NewRegistryValue(dataKey, sl.Bytes(), rev).Sign(sk)
+	err = c.RegistryUpdate(spk, dataKey, srv.Revision, srv.Signature, sl)
+	if err != nil {
+		return skymodules.Skylink{}, err
+	}
+	return skymodules.NewSkylinkV2(spk, dataKey), nil
+}
+
+// registryRead reads a registry entry and returns the SkylinkV2 it contains, as well
+// as the revision.
+func registryRead(c *client.Client, pk crypto.PublicKey, dataKey crypto.Hash) (skymodules.Skylink, uint64, error) {
+	spk := types.SiaPublicKey{
+		Algorithm: types.SignatureEd25519,
+		Key:       pk[:],
+	}
+
+	srv, err := c.RegistryRead(spk, dataKey)
+	if err != nil {
+		return skymodules.Skylink{}, 0, errors.AddContext(err, "failed to read from the registry")
+	}
+	err = srv.Verify(pk)
+	if err != nil {
+		return skymodules.Skylink{}, 0, errors.AddContext(err, "the value we read failed validation")
+	}
+	var sl skymodules.Skylink
+	err = sl.LoadBytes(srv.Data)
+	if err != nil {
+		return skymodules.Skylink{}, 0, errors.AddContext(err, "registry value is not a SkylinkV2")
+	}
+	return sl, srv.Revision, nil
+}
+
+// uploadData uploads the given data to skynet and returns a SkylinkV1.
+func uploadData(c *client.Client, content []byte) (string, error) {
+	buf := bytes.NewReader(content)
+	fNameBytes := sha1.Sum(content)
+	fName := base64.StdEncoding.EncodeToString(fNameBytes[:])
+	if strings.HasSuffix(fName, "/") {
+		fName = fName[1:]
+	}
+	sp, err := skymodules.NewSiaPath(fName)
+	if err != nil {
+		return "", errors.AddContext(err, "failed to create new sia path")
+	}
+	sup := &skymodules.SkyfileUploadParameters{
+		SiaPath:  sp,
+		Filename: "cert.json", // fName
```

Comment:
turns out it's not always certificate there, just making it more generic 
```suggestion
		Filename: "data.json", // fName
```